src/Pos/Explorer/BListener.hs

```diff
@@ -0,0 +1,255 @@
+onApplyEpochBlocksExplorer blunds = do
+    -- Get the epoch and the attached block @HeaderHash@es.
+    let epoch    = epochBlocks ^. _1
+    let blockHHs = epochBlocks ^. _2
+    
+    -- Get all blocks from the current epoch (so we can remove everything and
+    -- add the (Epoch, [HeaderHash]) pair back to the database).
+    mBlocksHHs <- DB.getEpochBlocks epoch
+    let blocksHHsOrEmpty = fromMaybe mempty mBlocksHHs
+    let uniqueHHs = ordNub $ blocksHHsOrEmpty <> blockHHs
+
+    -- In the end make sure we place this under @SomeBatchOp@ in order to
+    -- preserve atomicity.
+    pure $ SomeBatchOp $ DB.PutEpochBlocks epoch uniqueHHs
+  where
+    epochBlocks = minEpochBlocks blocksNE
+
+    blocksNE :: NE (Block ssc)  
+    blocksNE = fst <$> getOldestFirst blunds
+
+
+-- For @PageBlocks@
+onApplyPageBlocksExplorer
+    :: forall ssc m .
+    ( SscHelpersClass ssc
+    , WithLogger m
+    , MonadRealDB m
+    , MonadDBRead m
+    )
+    => OldestFirst NE (Blund ssc) -> m SomeBatchOp
+onApplyPageBlocksExplorer blunds = do
+    -- [(Integer, [HeaderHash])]
+    let newPageBlocks = blocksPagePages blocksNE
+    let pages = fst <$> newPageBlocks
+    -- Get existing @HeaderHash@es from the pages so we can merge them.
+    existingPageBlocks <- sequence [ getPagePageBlocks page | page <- pages]
+
+    -- convert to Map
+    let newPageBlocksMap      = M.fromList newPageBlocks
+    let existingPageBlocksMap = M.fromList existingPageBlocks
+
+    -- Merge the old and the new
+    let newExistingPageBlocks = M.toList $ M.unionWith (<>) newPageBlocksMap existingPageBlocksMap
+
+    let deletedPages = [ DB.DelPageBlocks page | page <- pages]
+    let createdPages = [ DB.PutPageBlocks (pB ^. _1) (ordNub $ pB ^. _2) | pB <- newExistingPageBlocks]
+
+    -- In the end make sure we place this under @SomeBatchOp@ in order to
+    -- preserve atomicity.
+    pure $ SomeBatchOp [deletedPages, createdPages]
+  where
+    blocksNE :: NE (Block ssc)  
+    blocksNE = fst <$> getOldestFirst blunds
+
+    -- Get exisiting page blocks paired with the page number. If there are no
+    -- saved blocks on the page return an empty list.
+    getPagePageBlocks 
+        :: (MonadDBRead m) 
+        => Page 
+        -> m (Page, [HeaderHash])
+    getPagePageBlocks page = do
+        pageBlocks      <- DB.getPageBlocks page
+        let mPageBlocks = fromMaybe [] pageBlocks
+        pure (page, mPageBlocks)
+
+----------------------------------------------------------------------------
+-- Rollback
+----------------------------------------------------------------------------
+
+-- For @EpochBlocks@
+onRollbackEpochBlocksExplorer
+    :: forall ssc m .
+    ( SscHelpersClass ssc
+    , WithLogger m
+    , MonadRealDB m
+    , MonadDBRead m
+    )
+    => NewestFirst NE (Blund ssc) -> m SomeBatchOp
+onRollbackEpochBlocksExplorer blunds = do
+    let epoch    = epochBlocks ^. _1
+    pure $ SomeBatchOp $ DB.DelEpochBlocks epoch
```

Comment:
Wait, what is it? It's completely wrong. For example, suppose we have 4 blocks for epoch 3. And we rollback only one of these blocks. You will delete whole index for epoch 3. But you should remove only one element. You can get whole list, remove one item from there and put it back (or delete if it becomes empty).